File: mobile/test/runtime_contract_manifest_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:torchat_mobile/core/models/domain.dart';
import 'package:torchat_mobile/core/runtime/runtime_contract.dart';

Map<String, dynamic> manifest() => Map<String, dynamic>.from(
  jsonDecode(File('../common/client-engine-contract.json').readAsStringSync())
      as Map,
);

List<String> strings(Object? value) =>
    (value as List).map((item) => item.toString()).toList(growable: false);

void main() {
  test('Flutter EngineContract matches shared manifest', () {
    final contract = manifest();
    final methods = Map<String, dynamic>.from(contract['methods'] as Map);

    expect(strings(methods['public']), [
      EngineContract.bootstrap,
      EngineContract.connect,
      EngineContract.getIdentity,
      EngineContract.getProfile,
      EngineContract.pairingInbox,
      EngineContract.pairingOutbox,
      EngineContract.listContacts,
      EngineContract.listConversations,
      EngineContract.listMessages,
      EngineContract.getPeerEndpoint,
      EngineContract.retryPeerConnection,
      EngineContract.rotatePeerEndpoint,
      EngineContract.setNickname,
      EngineContract.refreshPairingCode,
      EngineContract.submitPairingCode,
      EngineContract.acceptPairing,
      EngineContract.rejectPairing,
      EngineContract.archivePairing,
      EngineContract.cancelPairing,
      EngineContract.verifyContact,
      EngineContract.updateContactSettings,
      EngineContract.startConversation,
      EngineContract.openConversation,
      EngineContract.closeConversation,
      EngineContract.sendMessage,
      EngineContract.retryMessage,
      EngineContract.deleteMessageLocal,
      EngineContract.setTyping,
      EngineContract.setPresence,
      EngineContract.sendReadReceipts,
      EngineContract.platformFact,
      EngineContract.shutdown,
    ]);
    expect(strings(contract['events']), [
      EngineContract.runtimeReady,
      EngineContract.torStatus,
      EngineContract.profileReady,
      EngineContract.inviteReceived,
      EngineContract.inviteStateChanged,
      EngineContract.messageReceived,
      EngineContract.messageStateChanged,
      EngineContract.conversationReadChanged,
      EngineContract.typingChanged,
      EngineContract.presenceChanged,
      EngineContract.peerEndpointChanged,
      EngineContract.peerConnectionChanged,
      EngineContract.changed,
      EngineContract.runtimeError,
      EngineContract.runtimeLog,
    ]);
  });

  test('Flutter domain parsers accept only canonical manifest enum values', () {
    final contract = manifest();

    for (final value in strings(contract['messageStates'])) {
      expect(MessageState.fromValue(value).wireValue, value);
    }
    for (final value in strings(contract['conversationStates'])) {
      expect(
        ConversationSummary.fromMap({
          'id': 'c',
          'contactInstallationId': 'peer',
          'status': value,
        }).state.wireValue,
        value,
      );
    }
    for (final value in strings(contract['inviteStates'])) {
      expect(InviteState.fromValue(value).wireValue, value);
    }
    for (final value in strings(contract['pairingAvailableActions'])) {
      expect(PairingAvailableAction.fromValue(value).wireValue, value);
    }

    expect(() => MessageState.fromValue('PENDING'), throwsFormatException);
    expect(
      () => ConversationSummary.fromMap(const {
        'id': 'c',
        'contactInstallationId': 'peer',
        'status': 'NEW',
      }),
      throwsFormatException,
    );
    expect(() => InviteState.fromValue('CANCELED'), throwsFormatException);
  });
}

File: mobile/android/app/src/test/kotlin/org/torchat/mobile/RuntimeContractManifestTest.kt
package org.torchat.mobile

import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Test
import org.torchat.generated.EngineContract
import java.io.File

class RuntimeContractManifestTest {
    private fun manifest(): JSONObject {
        val candidates = listOf(
            File("../../common/client-engine-contract.json"),
            File("../../../common/client-engine-contract.json"),
        )
        val file = candidates.firstOrNull { it.isFile }
            ?: error("common/client-engine-contract.json not found from ${File(".").absolutePath}")
        return JSONObject(file.readText())
    }

    private fun strings(json: JSONObject, key: String): List<String> {
        val array = json.getJSONArray(key)
        return (0 until array.length()).map { index -> array.getString(index) }
    }

    @Test
    fun runtime_contract_constants_match_shared_manifest() {
        val contract = manifest()
        val methods = contract.getJSONObject("methods")

        assertEquals(
            listOf(
                EngineContract.BOOTSTRAP,
                EngineContract.CONNECT,
                EngineContract.GET_IDENTITY,
                EngineContract.GET_PROFILE,
                EngineContract.PAIRING_INBOX,
                EngineContract.PAIRING_OUTBOX,
                EngineContract.LIST_CONTACTS,
                EngineContract.LIST_CONVERSATIONS,
                EngineContract.LIST_MESSAGES,
                EngineContract.SET_NICKNAME,
                EngineContract.REFRESH_PAIRING_CODE,
                EngineContract.SUBMIT_PAIRING_CODE,
                EngineContract.ACCEPT_PAIRING,
                EngineContract.REJECT_PAIRING,
                EngineContract.ARCHIVE_PAIRING,
                EngineContract.CANCEL_PAIRING,
                EngineContract.VERIFY_CONTACT,
                EngineContract.START_CONVERSATION,
                EngineContract.OPEN_CONVERSATION,
                EngineContract.CLOSE_CONVERSATION,
                EngineContract.SEND_MESSAGE,
                EngineContract.PLATFORM_FACT,
                EngineContract.SHUTDOWN,
            ),
            strings(methods, "public"),
        )
        assertEquals(
            listOf(
                EngineContract.RUNTIME_READY,
                EngineContract.TOR_STATUS,
                EngineContract.PROFILE_READY,
                EngineContract.INVITE_RECEIVED,
                EngineContract.INVITE_STATE_CHANGED,
                EngineContract.MESSAGE_RECEIVED,
                EngineContract.MESSAGE_STATE_CHANGED,
                EngineContract.CONVERSATION_READ_CHANGED,
                EngineContract.CHANGED,
                EngineContract.RUNTIME_ERROR,
                EngineContract.RUNTIME_LOG,
            ),
            strings(contract, "events"),
        )
    }

    @Test
    fun canonical_transport_and_pairing_outcomes_match_manifest() {
        val contract = manifest()

        assertEquals(
            listOf(
                EngineContract.OUTCOME_FORWARDED,
                EngineContract.OUTCOME_DELIVERED,
                EngineContract.OUTCOME_RECIPIENT_OFFLINE,
                EngineContract.OUTCOME_RETRYABLE_FAILURE,
                EngineContract.OUTCOME_PERMANENT_FAILURE,
            ),
            strings(contract, "messageTransportOutcomes"),
        )
        assertEquals(
            listOf(
                EngineContract.PAIRING_OUTCOME_OFFER_RECEIVED,
                EngineContract.PAIRING_OUTCOME_REJECTION_RECEIVED,
                EngineContract.PAIRING_OUTCOME_WELCOME_PREPARED,
            ),
            strings(contract, "pairingPeerOutcomes"),
        )
    }
}
